fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb

```diff
@@ -80,36 +80,56 @@ def run!
         display_user_error!(e, e.to_s)
       rescue FastlaneCore::Interface::FastlaneError => e # user_error!
-        collector.did_raise_error(@program[:name])
-        show_github_issues(e.message) if e.show_github_issues
-        FastlaneCore::CrashReporter.report_crash(exception: e, action: @program[:name])
-        display_user_error!(e, e.message)
+        rescue_fastlane_error(e)
       rescue Errno::ENOENT => e
-        # We're also printing the new-lines, as otherwise the message is not very visible in-between the error and the stacktrace
-        puts ""
-        FastlaneCore::UI.important("Error accessing file, this might be due to fastlane's directory handling")
-        FastlaneCore::UI.important("Check out https://docs.fastlane.tools/advanced/#directory-behavior for more details")
-        puts ""
-        FastlaneCore::CrashReporter.report_crash(exception: e, action: @program[:name])
-        raise e
+        rescue_file_error(e)
       rescue Faraday::SSLError => e # SSL issues are very common
         handle_ssl_error!(e)
       rescue Faraday::ConnectionFailed => e
-        if e.message.include? 'Connection reset by peer - SSL_connect'
-          handle_tls_error!(e)
-        else
-          FastlaneCore::CrashReporter.report_crash(exception: e, action: @program[:name])
-          handle_unknown_error!(e)
-        end
+        rescue_connection_failed_error(e)
       rescue => e # high chance this is actually FastlaneCore::Interface::FastlaneCrash, but can be anything else
+        rescue_unknown_error(e)
+      ensure
+        @collector.did_finish
+      end
+    end
+
+    def rescue_file_error(e)
+      # We're also printing the new-lines, as otherwise the message is not very visible in-between the error and the stacktrace
+      puts ""
+      FastlaneCore::UI.important("Error accessing file, this might be due to fastlane's directory handling")
+      FastlaneCore::UI.important("Check out https://docs.fastlane.tools/advanced/#directory-behavior for more details")
+      puts ""
+      FastlaneCore::CrashReporter.report_crash(exception: e, action: @program[:name])
+      raise e
+    end
+
+    def rescue_connection_failed_error(e)
+      if e.message.include? 'Connection reset by peer - SSL_connect'
+        handle_tls_error!(e)
+      else
         FastlaneCore::CrashReporter.report_crash(exception: e, action: @program[:name])
-        collector.did_crash(@program[:name])
         handle_unknown_error!(e)
-      ensure
-        collector.did_finish
       end
     end
 
+    def rescue_unknown_error(e)
+      FastlaneCore::CrashReporter.report_crash(exception: e, action: @program[:name])
+      unless e.fastlane_crash_came_from_custom_action? || e.fastlane_crash_came_from_plugin?
+        @collector.did_crash(@program[:name])
+      end
+      handle_unknown_error!(e)
+    end
+
+    def rescue_fastlane_error(e)
+      unless e.fastlane_crash_came_from_custom_action? || e.fastlane_crash_came_from_plugin?
```

Comment:
Oh I just noticed that we have `fastlane_crash_came_from_custom_action? || e.fastlane_crash_came_from_plugin` 4 times in total, maybe we can add another helper method that would unify those 2 to make the rest of the code base easier to read?